### PR TITLE
Update DMAPI to version 3.2.1.0

### PR DESCRIPTION
**Pull request by Cyberboss**
This is primarily a compatibility layer that enables both forward and backward compatibility for all past and future APIs based on the detected running TGS version. It also bundles all it's includes to make future upgrades not have to modify the dme, heck, could even use a submodule if it wanted (not happening). No other changes necessary.

There's an upcoming event system and new chat management functions. Check them out here: https://github.com/tgstation/tgstation-server/blob/303448457e1529c900fa183a8b3c7f2e5459acf1/DMAPI/tgs.dm

Also added /datum/proc/CanProcCall()

### CODE REVIEW

File: code/__DEFINES/tgs.dm [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+//REQUIRED HOOKS
+
+//Call this somewhere in /world/New() that is always run
+//event_handler: optional user defined event handler. The default behaviour is to broadcast the event in english to all connected admin channels
+/world/proc/TgsNew(datum/tgs_event_handler/event_handler)
+	return
+
+//Call this when your initializations are complete and your game is ready to play before any player interactions happen
+//This may use world.sleep_offline to make this happen so ensure no changes are made to it while this call is running
+/world/proc/TgsInitializationComplete()
```
Comment (Cyberboss) [inline code/__DEFINES/tgs.dm]: @MrStonedOne

File: code/controllers/master.dm [hunk 2]
```diff
@@ -199,11 +199,14 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
...
 	// Set world options.
-	if(sleep_offline_after_initializations)
-		world.sleep_offline = TRUE
 	world.fps = CONFIG_GET(number/fps)
 	var/initialized_tod = REALTIMEOFDAY
+
+	world.TgsInitializationComplete()
+	if(sleep_offline_after_initializations)
+		world.sleep_offline = TRUE
 	sleep(1)
```
Comment (Cyberboss) [inline code/controllers/master.dm]: I put it here, don't know if you want it anywhere else. It's unimplemented in v3 but, like you suggest, the plan is to start the world in the background and then transfer the players to it when the current round ends. If you want it somewhere else speak now or forever hold your peace.

**Comment by optimumtact** [comment]
```
[20:59:52] Runtime in core.dm, line 23: list index out of bounds
proc name: SelectTgsApi (/world/proc/SelectTgsApi)
src: world
call stack:
world: SelectTgsApi("3.2.2.0")
world: TgsNew(null)
world: New()

[20:59:52] Runtime in core.dm, line 10: Cannot create objects of type null.
proc name: TgsNew (/world/TgsNew)
src: world
call stack:
world: TgsNew(null)
world: New()
```

File: code/modules/tgs/core/core.dm [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	if(result && result != TGS_UNIMPLEMENTED)
+		TGS_WRITE_GLOBAL(tgs, new_api)
+
+/world/proc/SelectTgsApi(tgs_version)
+	//remove the old 3.0 header
+	tgs_version = replacetext(tgs_version, "/tg/station 13 Server v", "")
+
+	var/list/version_bits = splittext(tgs_version, ".")
+
+	var/super = text2num(version_bits[0])
```
Comment (MrStonedOne) [inline code/modules/tgs/core/core.dm]: >0

hmmmmmmmmmmmmmmmmmmmm

Comment (optimumtact) [inline code/modules/tgs/core/core.dm]: :thinking:

**Comment by Cyberboss** [comment]
FUCK

**Comment by Cyberboss** [comment]
This is ready

**Pull request merged**